<a href="https://colab.research.google.com/github/abm4github/cardiovascular-risk-prediction_Supervised-Classification/blob/main/AlmaBetter_ML_Classification_CardioVascularRisk.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 

In [5]:
# mount drive and give path of dataset
data = pd.read_csv("/content/drive/MyDrive/AlmaBetter ML Classification Deliverables/data_cardiovascular_risk.csv")

In [6]:
data.sample(5)

id  age  education sex  ...    BMI  heartRate  glucose  TenYearCHD
3179  3179   40        4.0   M  ...  35.20       66.0     55.0           0
3070  3070   35        3.0   M  ...  21.03       73.0     69.0           1
1433  1433   55        1.0   F  ...  30.61       72.0     73.0           0
2425  2425   62        1.0   F  ...  32.19       94.0    117.0           0
1669  1669   61        2.0   F  ...  38.46       57.0     78.0           0

[5 rows x 17 columns]

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3390 entries, 0 to 3389
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               3390 non-null   int64  
 1   age              3390 non-null   int64  
 2   education        3303 non-null   float64
 3   sex              3390 non-null   object 
 4   is_smoking       3390 non-null   object 
 5   cigsPerDay       3368 non-null   float64
 6   BPMeds           3346 non-null   float64
 7   prevalentStroke  3390 non-null   int64  
 8   prevalentHyp     3390 non-null   int64  
 9   diabetes         3390 non-null   int64  
 10  totChol          3352 non-null   float64
 11  sysBP            3390 non-null   float64
 12  diaBP            3390 non-null   float64
 13  BMI              3376 non-null   float64
 14  heartRate        3389 non-null   float64
 15  glucose          3086 non-null   float64
 16  TenYearCHD       3390 non-null   int64  
dtypes: float64(9),

In [9]:
data.isna().sum()

id                   0
age                  0
education           87
sex                  0
is_smoking           0
cigsPerDay          22
BPMeds              44
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             38
sysBP                0
diaBP                0
BMI                 14
heartRate            1
glucose            304
TenYearCHD           0
dtype: int64

##Dealing with null values

Note: This part is postponed for later stages, now proceeding with deletion of null values.

Missing values in variable "glucose" are 304 in count, deleting them may cause loss of data, so checking correlation between "glucose" and "TenYearCHD"

In [12]:
# glucose_TenYearCHD_data = data[['glucose','TenYearCHD']]
# correlation_1 = glucose_TenYearCHD_data.corr(method = 'pearson')
# print(correlation_1)

            glucose  TenYearCHD
glucose      1.0000      0.1382
TenYearCHD   0.1382      1.0000


In [13]:
# data['glucose'].describe()

count    3086.000000
mean       82.086520
std        24.244753
min        40.000000
25%        71.000000
50%        78.000000
75%        87.000000
max       394.000000
Name: glucose, dtype: float64

Deleting rows with Null Values

In [14]:
data.dropna(inplace=True)

Re-checking null values, if any and new shape of data

In [15]:
data.isna().sum().sum()

0

In [18]:
data.shape

(2927, 17)

##Dropping irrelavant variables and duplicates if any.

In [16]:
data.columns

Index(['id', 'age', 'education', 'sex', 'is_smoking', 'cigsPerDay', 'BPMeds',
       'prevalentStroke', 'prevalentHyp', 'diabetes', 'totChol', 'sysBP',
       'diaBP', 'BMI', 'heartRate', 'glucose', 'TenYearCHD'],
      dtype='object')

In [19]:
data.drop("id", axis=1, inplace=True)
data.drop_duplicates(inplace=True)

In [20]:
data.shape

(2927, 16)

here shape confirms that there are no duplicates and total columns reduced to 16

In [22]:
data.describe()

age    education  ...      glucose   TenYearCHD
count  2927.000000  2927.000000  ...  2927.000000  2927.000000
mean     49.507345     1.965494  ...    81.934404     0.151691
std       8.597191     1.017199  ...    24.106943     0.358783
min      32.000000     1.000000  ...    40.000000     0.000000
25%      42.000000     1.000000  ...    71.000000     0.000000
50%      49.000000     2.000000  ...    78.000000     0.000000
75%      56.000000     3.000000  ...    87.000000     0.000000
max      70.000000     4.000000  ...   394.000000     1.000000

[8 rows x 14 columns]

##Dealing with object type variables 

In [24]:
data['sex'].unique()

array(['M', 'F'], dtype=object)

In [25]:
data['is_smoking'].unique()

array(['NO', 'YES'], dtype=object)

In [27]:
data = pd.get_dummies(data, drop_first=True)

In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2927 entries, 1 to 3389
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   age              2927 non-null   int64  
 1   education        2927 non-null   float64
 2   cigsPerDay       2927 non-null   float64
 3   BPMeds           2927 non-null   float64
 4   prevalentStroke  2927 non-null   int64  
 5   prevalentHyp     2927 non-null   int64  
 6   diabetes         2927 non-null   int64  
 7   totChol          2927 non-null   float64
 8   sysBP            2927 non-null   float64
 9   diaBP            2927 non-null   float64
 10  BMI              2927 non-null   float64
 11  heartRate        2927 non-null   float64
 12  glucose          2927 non-null   float64
 13  TenYearCHD       2927 non-null   int64  
 14  sex_M            2927 non-null   uint8  
 15  is_smoking_YES   2927 non-null   uint8  
dtypes: float64(9), int64(5), uint8(2)
memory usage: 348.7 KB


our data is now free of object type data

## Data prepation to modeling

In [32]:
data.sample(5)

age  education  cigsPerDay  ...  TenYearCHD  sex_M  is_smoking_YES
3174   39        3.0         0.0  ...           0      0               0
3374   67        2.0        60.0  ...           1      1               1
2181   60        1.0         0.0  ...           1      1               0
1614   50        2.0        20.0  ...           0      0               1
1224   52        1.0         0.0  ...           0      0               0

[5 rows x 16 columns]

Splitting target variable from independent variables

In [30]:
target_variable = 'TenYearCHD'
X = data.drop([target_variable], axis=1).values
y = data[target_variable].values

In [31]:
X

array([[36.,  4.,  0., ..., 75.,  1.,  0.],
       [46.,  1., 10., ..., 94.,  0.,  1.],
       [50.,  1., 20., ..., 94.,  1.,  1.],
       ...,
       [46.,  1.,  0., ..., 84.,  0.,  0.],
       [44.,  3.,  3., ..., 72.,  1.,  1.],
       [54.,  3.,  0., ..., 92.,  0.,  0.]])

In [33]:
y

array([0, 0, 1, ..., 0, 1, 0])

Splitting train and test data

In [34]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [40]:
X_train

array([[53.,  2., 20., ..., 80.,  0.,  1.],
       [53.,  1.,  0., ..., 77.,  0.,  0.],
       [57.,  1., 20., ..., 83.,  1.,  1.],
       ...,
       [46.,  2., 15., ..., 86.,  0.,  1.],
       [68.,  1.,  0., ..., 79.,  1.,  0.],
       [39.,  3.,  0., ..., 88.,  0.,  0.]])

In [36]:
y_train

array([0, 0, 1, ..., 0, 1, 0])

In [37]:
X_test

array([[65.,  1.,  3., ..., 78.,  0.,  1.],
       [56.,  2.,  5., ..., 85.,  0.,  1.],
       [56.,  3.,  0., ..., 75.,  0.,  0.],
       ...,
       [42.,  1.,  0., ..., 96.,  0.,  0.],
       [64.,  4.,  9., ..., 62.,  0.,  1.],
       [50.,  2.,  0., ..., 90.,  0.,  0.]])

In [43]:
y_test

array([0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,

Feature Scaling X_train and X_test

In [44]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_scaled = sc.fit_transform(X_train)
X_test_scaled = sc.transform(X_test)

In [45]:
X_train_scaled

array([[ 0.41690903,  0.03194675,  0.92066157, ..., -0.06742334,
        -0.89563134,  1.01246556],
       [ 0.41690903, -0.95209728, -0.76255037, ..., -0.20396143,
        -0.89563134, -0.98768792],
       [ 0.88061448, -0.95209728,  0.92066157, ...,  0.06911475,
         1.11653082,  1.01246556],
       ...,
       [-0.3945755 ,  0.03194675,  0.49985858, ...,  0.20565285,
        -0.89563134,  1.01246556],
       [ 2.15580445, -0.95209728, -0.76255037, ..., -0.11293604,
         1.11653082, -0.98768792],
       [-1.20606003,  1.01599078, -0.76255037, ...,  0.29667824,
        -0.89563134, -0.98768792]])

In [46]:
X_test_scaled

array([[ 1.80802537, -0.95209728, -0.51006858, ..., -0.15844873,
        -0.89563134,  1.01246556],
       [ 0.76468811,  0.03194675, -0.34174739, ...,  0.16014015,
        -0.89563134,  1.01246556],
       [ 0.76468811,  1.01599078, -0.76255037, ..., -0.29498683,
        -0.89563134, -0.98768792],
       ...,
       [-0.85828095, -0.95209728, -0.76255037, ...,  0.66077982,
        -0.89563134, -0.98768792],
       [ 1.69209901,  2.00003481, -0.005105  , ..., -0.88665189,
        -0.89563134,  1.01246556],
       [ 0.06912994,  0.03194675, -0.76255037, ...,  0.38770364,
        -0.89563134, -0.98768792]])

Here is a data ready to be trained and tested, using any machine learning classification algorithm. 

## Decision Tree Classification

Training the Decision Tree Classification model on the Training set

In [47]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train_scaled, y_train)

DecisionTreeClassifier(criterion='entropy', random_state=0)

Making the Confusion matrix

In [48]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[420  72]
 [ 64  30]]


0.7679180887372014

##K-Nearest Neighbors (K-NN)

Training the K-NN model on the Training set

In [49]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
classifier.fit(X_train_scaled, y_train)

KNeighborsClassifier()

Making the Confusion Matrix

In [50]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[473  19]
 [ 87   7]]


0.8191126279863481

##Kernel SVM

Training the Kernel SVM model on the Training set

In [51]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train_scaled, y_train)

SVC(random_state=0)

Making the Confusion Matrix

In [52]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[491   1]
 [ 94   0]]


0.8378839590443686

##Logistic Regression

Training the Logistic Regression model on the Training set

In [53]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train_scaled, y_train)

LogisticRegression(random_state=0)

Making the Confusion Matrix

In [54]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[485   7]
 [ 87   7]]


0.8395904436860068

##Naive Bayes

Training the Naive Bayes model on the Training set

In [58]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train_scaled, y_train)

GaussianNB()

Making the Confusion Matrix

In [59]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[452  40]
 [ 72  22]]


0.8088737201365188

##Random Forest Classification

Training the Random Forest Classification model on the Training set

In [60]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train_scaled, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)

Making the Confusion Matrix

In [61]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[477  15]
 [ 91   3]]


0.8191126279863481

##Support Vector Machine (SVM)

Training the SVM model on the Training set

In [62]:
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train_scaled, y_train)

SVC(kernel='linear', random_state=0)

Making the Confusion Matrix

In [63]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test_scaled)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[492   0]
 [ 94   0]]


0.8395904436860068

##Accuracy Scores of each classification model are as follows 

In [66]:
print('''Accuracy Scores:

Decision Tree Classification: 0.7679180887372014
K Nearest Neigbours         : 0.8191126279863481
Kernal SVM                  : 0.8378839590443686
Logistic Regression         : 0.8395904436860068
Naive Bayes                 : 0.8088737201365188
Random Forest Classification: 0.8191126279863481
Support Vector Machine      : 0.8395904436860068
''')

Accuracy Scores:

Decision Tree Classification: 0.7679180887372014
K Nearest Neigbours         : 0.8191126279863481
Kernal SVM                  : 0.8378839590443686
Logistic Regression         : 0.8395904436860068
Naive Bayes                 : 0.8088737201365188
Random Forest Classification: 0.8191126279863481
Support Vector Machine      : 0.8395904436860068

